.github/workflows/go.yml

```diff
@@ -0,0 +1,155 @@
+name: Build and Test
+
+on:
+  push:
+    branches:
+      - '*'
+  pull_request:
+    branches:
+      - '*'
+
+jobs:
+  build-and-test:
+    name: Build Go and Test
+    strategy:
+      matrix:
+        os: [ubuntu-latest, macOS-latest, windows-latest]
+    runs-on: ${{ matrix.os }}
+    steps:
+      - name: Set up Go
+        uses: actions/setup-go@v1
+        with:
+          go-version: 1.13
+        id: go
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Check out code into the Go module directory
+        uses: actions/checkout@master
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+        # Needed for changelog generation by goreleaser
+      - name: Unshallow
+        run: git fetch --prune --unshallow
+
+      - name: Get dependencies
+        run: make deps
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+
+      - name: Test (Linux/macOS)
+        if: matrix.os != 'windows-latest'
+        run: make coverage
+        timeout-minutes: 3
+        env:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+          SYSL_PLANTUML: http://www.plantuml.com/plantuml
+
+      - name: Test (Windows)
+        if: matrix.os == 'windows-latest'
```

Comment:
Making these all steps instead of jobs means they will all run in series. This is going to slow down the build and make it take longer to merge PRs. I think you should at least keep windows as a separate job as it shares no steps with the unix-like build.

I'm not sure how the matrix interacts with steps, but if every step has a !windows or windows gate, there's little point putting it into a matrix.

Reply:
ok, so I read up on the github build matrix, and each runs as a separate job, so I guess my objection is moot, but this still reads awkwardly with all the intermixed OS steps. Kind of makes a mockery of the matrix stuff. I wonder if there's a better way - could we include the common steps into other jobs somehow, or parameterise the os stuff in a makefile or something? thinking out aloud.

Reply:
Yup, in this case, they do run in parallel. I agree with the readability part though, it suffers cause of the branching. I’ll take a look and see if there’s ant other way but I think the reduction in duplication here is worth it

Reply:
The duplication can be reduced to:
```
      - name: Set up Go
        uses: actions/setup-go@v1
        with:
          go-version: 1.13
      - name: Check out code into the Go module directory
        uses: actions/checkout@master
        # Needed for changelog generation by goreleaser
      - name: Unshallow
        run: git fetch --prune --unshallow
```

This also implies that the `go mod tidy` (`make deps`) check is done in a different job (without the unshallow as it is not needed). I think a different job makes sense for that as it is a separate gate. Also the `make deps` part is not really needed. Each step will download the deps it needs as it builds, so you don't need the `go get ...` stuff explicitly.

Reply:
Also, I think https://github.com/actions/checkout#fetch-all-history-for-all-tags-and-branches is saying you can do the unshallow in the same step as the checkout.

Reply:
Had a look at the options, and it looks like there's 2 options for making it more readable:
1. Put all the conditional steps into a makefile (would make the workflow more readable, as the workflow would just do the sequencing, and the makefile provides the actual task definition)
2. Put the common steps in their own workflow, and maintain the 3 separate workflow files (This seems really messy, since you'd end up with multiple workflow files, might not work since subsequent steps are dependant on previous ones. You'd have to share the files/ artifacts in some way ) 

I'm learning towards option 1 more, but I'm hoping that I can do that in a seperate PR, along with a more comprehensive cleanup of the Makefile. Thoughts?

Reply:
do they need to be separate workflow files? - there is already two jobs in the one file here. can't you do the same with the build/test jobs?

Reply:
hmm, I think you can run them as dependant jobs, so they don't run in parallel, but you'd still have similar difficulties with sharing files etc, as each job runs in its own environment.

Definitely seeing the limitations of trying to build a workflow here with a YAML based DSL here!

Reply:
i dont think there is any sharing of files across jobs whether you use a matrix or not.